### DISCUSSION TITLE
Inappbrowser cordova plugin deployment issue

### DISCUSSION THREAD

**Discussion by zahra-alsabah** [apache/cordova#407] [Q&A]
I am using angular14 , cordova 11.0.0, and ios 6.2.0, I installed Inappbrowser cordova plugin and then built and it works perfectly on my local application. Afterward, I deployed it using Jenkins but the plugin doesn't work on the deployed version, I took the deployed project folder and tested it on Xcode, it shows this error :

typeerror: undefined is not an object (evaluating 'cordova.inappbrowser.open')

knowing that the installation and configuration are the same as the local one.

I tried :

1. removing and installing inappbrowser plugin\cordova and built afterward
2. building before and after updating the plugin
3. added `<plugin name="cordova-plugin-inappbrowser" />` and `<feature name="InAppBrowser"><param name="ios-package" value="CDVInAppBrowser" /></feature> `in config.xml

none of that worked

But when I removed the IOS platform and installed IOS 6.3.0 I made the plugin work locally, then deployed it on Jenkins but again it doesn't work on the deployed version

**Comment by breautek** [comment]
> added <plugin name="cordova-plugin-inappbrowser" /> and <feature name="InAppBrowser"><param name="ios-package" value="CDVInAppBrowser" /></feature> in config.xml

This shouldn't be necessary. `<plugin>` tags are deprecated/obsolete, the Cordova CLI will migrate them into `package.json`. The `<feature>` tag is something provided by the plugin itself.

> knowing that the installation and configuration are the same as the local one.

This might be what is believed, but I strongly suspect that something isn't the same and it might not be obvious.

Here are some recommendations, broken down in two main parts..

#### First, make sure your project setup is reproducible.

What I mean by this is you should be able to git clone (or whatever what your version control system is) and pull down a fresh checkout of your cordova project. If you do not use a version control system, I'd strongly recommend using one. For the rest of this post, I'd be assuming you're using git. I'll also assume that all generated files are in your  `.gitignore` in your version control system, including:

- platforms/
- plugins/
- node_modules/

So given those assumptions, at minimum the steps to checkout and build/run a project would be:

```
git clone myGitCorodvaProjectUrl
cd myGitCorodvaProjectUrl
npm install
cordova platform add ios@6
cordova run ios
```

Note that to use `run` on the iOS platform, `ios-deploy` needs to be installed if I recall correctly. It's not included with Cordova because it's a GPL licensed product. Alternatively you can open up the `xcworkspace` project inside `<cordova-project>/platforms/ios/<your project name>.xcworkspace`. I'm not sure how to launch an app to device or simulator from command line otherwise.

If these steps do not produce an expected working build, then they definitely won't work on Jenkins. So the focus should be what is causing this flow to fail.

#### Now on jenkins,

If all above is good :+1:  then I'd recommend enabling the feature that always creates a fresh build environment for each task. I forget what the feature is called, it's been awhile since I used Jenkins. But there will be a configuration option that forces a fresh workspace for your task on every run. This will obviously be much slower as this means it will need to pull down your project, npm install and resetup all your plugins when you add your platform(s) (e.g. everything in step 1) but this flow should guarantee that the project state is identical.

If the problem only produces itself if you're reusing a workspace, then there might be a problem where something is modifying something that it shouldn't.

This might not answer your question/problem but hopefully these pointers will help get you on the right track.

**Reply by zahra-alsabah** [reply]
Thank you Norman for your response and the effort you made to explain the steps, I will keep you updated if that works or if any other method helped me
